### PR TITLE
Update plot layout

### PR DESCRIPTION
**Pull request by bishwobhandari**
- updated axes fix with confidence interval (used max stdevs for y max value and y min value). for dichotomous it uses [0,1]
- fixed legend or right for some datasets 
- when the axis are changed, it is stored for re rendering  the plot when model changes.

Axes settings preserved on resizing:
Screen Shot 2021-09-23 at 12 17 56 PM

Continuous examples:
Screen Shot 2021-09-23 at 1 06 51 PM
Screen Shot 2021-09-23 at 1 07 01 PM

Dichotomous examples:
Screen Shot 2021-09-23 at 12 18 07 PM
Screen Shot 2021-09-23 at 12 17 56 PM

### CODE REVIEW

**Comment by bishwobhandari** [comment]
@shapiromatron for the confidence interval, i just used max stdev for the dataset. so the y min and y max are calculated based on that for continuous model. or we need to calculate the confidence interval assuming 95%?

**Comment by shapiromatron** [comment]
> @shapiromatron for the confidence interval, i just used max stdev for the dataset. so the y min and y max are calculated based on that for continuous model. or we need to calculate the confidence interval assuming 95%?

We do something fancier for the error bars, especially in cases where the N is small it can often be larger than just the SD. https://github.com/shapiromatron/bmds-server/blob/main/frontend/src/utils/errorBars.js

The data should be available somewhere though. I'd rather use the calculated values if we have them, if not I'd probably play around with some example datasets to make sure the bounds work (maybe mean + 3*SD?)

@bishwobhandari let me know when you've had a chance to play with the bounds and I will test.

**Comment by shapiromatron** [comment]
> @shapiromatron I used the confidence interval value from the continuous error bars bound max value for the max y and min value for the min y. if I use max value for both max y and min y , the plot looks weird. please review.

Take a look at the bounds parameter. If i remember correctly, I dont think it's the absolute value but instead the diff from the mean value.  Am I remembering that right?

**Comment by shapiromatron** [comment]
Also, if a BMD session has two datasets, are the bound values effectively reset if a user toggles the active dataset view? @bishwobhandari

**Comment by bishwobhandari** [comment]
> Also, if a BMD session has two datasets, are the bound values effectively reset if a user toggles the active dataset view? @bishwobhandari

yes because the dataset would change for another view and the layout will also change based on dataset.

**Comment by bishwobhandari** [comment]
@shapiromatron ready for review.

File: frontend/src/components/IndividualModel/ModelDetailModal.js [hunk 1]
```diff
@@ -153,6 +153,7 @@ class ModelDetailModal extends Component {
 
         return (
             <Modal
+                id="contained-modal"
```
Comment (shapiromatron) [inline frontend/src/components/IndividualModel/ModelDetailModal.js]: this should not be here? I noticed a few others like this... this was a change from an unrelated PR. Please make sure your code is clean before submitting.

When starting a new issue, make sure you're pulling from the `main` branch when it's clean and synced with the origin. Then, you know only the changes you're making are regarding the issue at hand.

File: frontend/src/constants/plotting.js [hunk 2]
```diff
@@ -67,21 +66,38 @@ export const getDrLayout = function(dataset, selected, modal, hover) {
...
 
         layout.xaxis.range = [xmin == 0 ? -xbuff : xmin - xbuff, xmax == 0 ? xbuff : xmax + xbuff];
         layout.yaxis.range =
             dataset.dtype == Dtype.DICHOTOMOUS
-                ? [0, 1]
+                ? [-0.05, 1.05]
                 : [ymin == 0 ? -ybuff : ymin - ybuff, ymax == 0 ? ybuff : ymax + ybuff];
 
+        // determine whether to position legend to the left or right; auto doesn't work
+        const maxResponseIndex = response.indexOf(_.max(response)),
```
Comment (shapiromatron) [inline frontend/src/constants/plotting.js:90]: your variables here were using underscores, which is a python convention. I changed them to camelCase, which is javascript.

File: frontend/src/stores/OutputStore.js [hunk 3]
```diff
@@ -32,13 +32,15 @@ class OutputStore {
...
 
+    @observable userPlotSettings = {};
     @observable showInlineNotes = false;
     @action.bound toggleInlineNotes() {
         this.showInlineNotes = !this.showInlineNotes;
     }
 
     @action.bound setSelectedOutputIndex(output_id) {
         this.selectedOutputIndex = output_id;
+        this.userPlotSettings = {};
```
Comment (shapiromatron) [inline frontend/src/stores/OutputStore.js:43]: you have to reset if a user changes the output.

File: frontend/src/stores/OutputStore.js [hunk 4]
```diff
@@ -316,6 +325,22 @@ class OutputStore {
...
+
+    @action.bound saveUserPlotSettings(e) {
+        // set user configurable plot settings
+        if (_.has(e, "xaxis.range[0]") && _.has(e, "xaxis.range[1]")) {
+            this.userPlotSettings["xaxis"] = [e["xaxis.range[0]"], e["xaxis.range[1]"]];
+        }
+        if (_.has(e, "yaxis.range[0]") && _.has(e, "yaxis.range[1]")) {
+            this.userPlotSettings["yaxis"] = [e["yaxis.range[0]"], e["yaxis.range[1]"]];
+        }
+        if (_.has(e, "yaxis.autorange")) {
```
Comment (shapiromatron) [inline frontend/src/stores/OutputStore.js:337]: things got weird if you reset on plotly; i had to handle that case. Also, in some cases, only xaxis or yaxis values were set and not all. Please check all edge cases.

File: frontend/src/components/common/DoseResponsePlot.js [hunk 5]
```diff
@@ -9,14 +9,15 @@ import {getConfig} from "../../constants/plotting";
...
-        const {layout, data} = this.props;
+        const {layout, data, store} = this.props;
 
         return (
             <Plot
                 layout={toJS(layout)}
                 data={toJS(data)}
                 config={getConfig()}
                 style={{width: "100%"}}
+                onRelayout={e => store.saveUserPlotSettings(e)}
```
Comment (shapiromatron) [inline frontend/src/components/common/DoseResponsePlot.js]: no need to pass the whole store; just pass the relayout function to call. You also missed the `<DoseResponsePlot/>` on the modals, even though you marked the store as a required prop.

**Review by shapiromatron — APPROVED** [review]
I made a  number of changes. Please see my review comments.

**Comment by shapiromatron** [comment]
@bishwobhandari also, for visual changes like this, please add screenshots to the PR that document the changes that were made.

**Comment by shapiromatron** [comment]
> > Also, if a BMD session has two datasets, are the bound values effectively reset if a user toggles the active dataset view? @bishwobhandari
> 
> yes because the dataset would change for another view and the layout will also change based on dataset.

I didnt find this to be the case when I tested it, until I reset the state in `setSelectedOutputIndex`

**Pull request merged**